// @edx/paragon/dist/DataTable/selection/ControlledSelectHeader.js
const _excluded = ["rows"];
function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }
function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }
function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }
import React, { useContext, useMemo, useCallback } from 'react';
import PropTypes from 'prop-types';
import { CheckboxControl } from '../../Form';
import DataTableContext from '../DataTableContext';
import { clearPageSelectionAction, setSelectedRowsAction } from './data/actions';
function ControlledSelectHeader(_ref) {
  let {
      rows
    } = _ref,
    rest = _objectWithoutProperties(_ref, _excluded);
  const {
    itemCount,
    controlledTableSelections: [, dispatch],
    getToggleAllPageRowsSelectedProps,
    state: {
      selectedRowIds
    },
    isAllPageRowsSelected
  } = useContext(DataTableContext);
  const selectedPageRowIds = useMemo(() => Object.keys(selectedRowIds), [selectedRowIds]);
  const toggleAllPageRowsSelected = useCallback(() => {
    if (isAllPageRowsSelected) {
      dispatch(clearPageSelectionAction(selectedPageRowIds));
    } else {
      dispatch(setSelectedRowsAction(rows, itemCount));
    }
  }, [rows, selectedPageRowIds, isAllPageRowsSelected, dispatch, itemCount]);
  const toggleAllPageRowsSelectedProps = getToggleAllPageRowsSelectedProps();
  toggleAllPageRowsSelectedProps.isIndeterminate = toggleAllPageRowsSelectedProps.indeterminate;
  // delete unused ``indeterminate`` prop
  delete toggleAllPageRowsSelectedProps.indeterminate;
  return /*#__PURE__*/React.createElement("div", {
    className: "pgn__data-table__controlled-select"
  }, /*#__PURE__*/React.createElement(CheckboxControl, _extends({}, toggleAllPageRowsSelectedProps, {
    onChange: toggleAllPageRowsSelected
  }, rest)));
}
ControlledSelectHeader.propTypes = {
  rows: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired
  })).isRequired
};
export default ControlledSelectHeader;
